timeout command

PaulMndn opened the pull request:
timeout command implemented and tested.

closes #24

PaulMndn committed:
added TimeoutList, moved guild init to bot.on_ready() so that the event loop is running (needed in TimeoutList init)

PaulMndn committed:
added timeout command

PaulMndn committed:
Merge branch 'main' into feature/timeout

PaulMndn committed:
cleanup after merge and general cleanup

PaulMndn committed:
changed command name

PaulMndn committed:
improved UI for `scrim-timeout`

jaimevisser commented on scrimbot/timeoutlist.py:
```diff
@@ -0,0 +1,89 @@
+                               task))
+        super().__init__(l)
+    
+    def __contains__(self, item):
```
Having __contains__ work on id while the rest of the functions in this class work on the user tuple breaks the principle of least astonishment. https://en.wikipedia.org/wiki/Principle_of_least_astonishment

PaulMndn replied:
removed and rewritten

jaimevisser commented on scrimbot/timeoutlist.py:
```diff
@@ -0,0 +1,89 @@
+from datetime import datetime, timedelta, timezone
+
+
+class TimeoutList(list):
```
Having the class extend list gives it a lot of methods that need not be exposed. In this case composition over inheritance would be the better choice. https://en.wikipedia.org/wiki/Composition_over_inheritance

jaimevisser replied:
To clarify, the class that manages timeouts needs a list, but does not need to _be_ a list.

PaulMndn replied:
implemented the principle

jaimevisser commented on scrimbot/timeoutlist.py:
```diff
@@ -0,0 +1,89 @@
+        for u in users:
+            task = self.bot.loop.create_task(self._timeout_runout(u["user_id"]))
+            l.append(self.User(u["user_id"],
+                               datetime.fromisoformat(u["timeout"]),
```
All other stored times are using timestamps, would be more consistent to use those here as well.

PaulMndn replied:
Alright. My idea was to make it more humanly readable, but I'll change it.

jaimevisser reviewed (changes requested):
See inline comments.

PaulMndn committed:
changed according to review comments
- composition over inheritance
- removed __contains__()
- switched to storing time as timestamp not isoformat

PaulMndn committed:
Merge remote-tracking branch 'origin/main' into feature/timeout

# Resolved conflicts in:
#	scrimbot/__init__.py
#	scrimbot/guild.py

PaulMndn commented:
Thanks for the links btw :)

PaulMndn committed:
added Store for TimeoutList

PaulMndn committed:
minor cleanup

PaulMndn commented:
before you merge, I'm working on another change I'd like to include.
